File: mnheme_ports/cpp/mnheme.cpp
// mnheme/cpp/mnheme.cpp
#include "mnheme.hpp"
#include <algorithm>
#include <cassert>
#include <iomanip>
#include <random>
#include <set>
#include <sstream>

// ── Tiny SHA-256 via OpenSSL or manual ──────────────────────────
// For production use: link with -lcrypto (OpenSSL)
#ifdef _WIN32
  #include <wincrypt.h>
#else
  #include <openssl/sha.h>
#endif

namespace mnheme {

// ─── IndexEngine ────────────────────────────────────────────────

std::string IndexEngine::to_lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s;
}

std::set<std::string> IndexEngine::tokenize(const std::string& text) {
    std::set<std::string> tokens;
    std::string lower = to_lower(text);
    std::string token;
    for (char c : lower) {
        if (std::isalpha(static_cast<unsigned char>(c))) {
            token += c;
        } else {
            if (token.size() >= 3) tokens.insert(token);
            token.clear();
        }
    }
    if (token.size() >= 3) tokens.insert(token);
    return tokens;
}

void IndexEngine::index_record(uint64_t offset, const json& r) {
    std::string concept   = r.value("concept", "");
    std::string feeling   = r.value("feeling", "");
    std::string timestamp = r.value("timestamp", "");

    concept_[concept].push_back(offset);
    feeling_[feeling].push_back(offset);
    cf_[{concept, feeling}].push_back(offset);
    timeline_.push_back({timestamp, offset});
    all_.push_back(offset);

    if (r.contains("tags") && r["tags"].is_array()) {
        for (const auto& tag : r["tags"]) {
            std::string t = tag.get<std::string>();
            if (!t.empty()) tag_[t].push_back(offset);
        }
    }

    // Inverted full-text index
    std::string combined =
        r.value("content", "") + " " +
        r.value("concept", "") + " " +
        r.value("note",    "");
    for (const auto& tok : tokenize(combined)) {
        word_[tok].push_back(offset);
    }
}

size_t IndexEngine::rebuild(const std::vector<std::pair<uint64_t, json>>& records) {
    concept_.clear(); feeling_.clear(); tag_.clear();
    cf_.clear(); timeline_.clear(); all_.clear(); word_.clear();

    for (const auto& [offset, record] : records) {
        index_record(offset, record);
    }
    std::sort(timeline_.begin(), timeline_.end(),
              [](const auto& a, const auto& b){ return a.first < b.first; });
    return records.size();
}

std::vector<uint64_t> IndexEngine::offsets_by_concept(
    const std::string& concept,
    const std::string* feeling,
    bool oldest_first) const
{
    std::vector<uint64_t> offs;
    if (feeling) {
        auto it = cf_.find({concept, *feeling});
        if (it != cf_.end()) offs = it->second;
    } else {
        auto it = concept_.find(concept);
        if (it != concept_.end()) offs = it->second;
    }
    if (!oldest_first) std::reverse(offs.begin(), offs.end());
    return offs;
}

std::vector<uint64_t> IndexEngine::offsets_by_feeling(
    const std::string& feeling, bool oldest_first) const
{
    std::vector<uint64_t> offs;
    auto it = feeling_.find(feeling);
    if (it != feeling_.end()) offs = it->second;
    if (!oldest_first) std::reverse(offs.begin(), offs.end());
    return offs;
}

std::vector<uint64_t> IndexEngine::offsets_by_word(const std::string& word) const {
    std::string w = to_lower(word);
    auto it = word_.find(w);
    if (it == word_.end()) return {};
    auto offs = it->second;
    std::reverse(offs.begin(), offs.end());
    return offs;
}

std::vector<uint64_t> IndexEngine::all_offsets(bool oldest_first) const {
    auto offs = all_;
    if (!oldest_first) std::reverse(offs.begin(), offs.end());
    return offs;
}

std::vector<uint64_t> IndexEngine::timeline_offsets(const std::string& concept) const {
    std::set<uint64_t> cset;
    auto it = concept_.find(concept);
    if (it != concept_.end())
        cset.insert(it->second.begin(), it->second.end());
    std::vector<uint64_t> result;
    for (const auto& [ts, off] : timeline_)
        if (cset.count(off)) result.push_back(off);
    return result;
}

std::vector<std::string> IndexEngine::all_concepts() const {
    std::vector<std::string> v;
    for (const auto& [k, _] : concept_) v.push_back(k);
    std::sort(v.begin(), v.end());
    return v;
}

size_t IndexEngine::count(const std::string* concept, const std::string* feeling) const {
    if (concept && feeling) {
        auto it = cf_.find({*concept, *feeling});
        return it == cf_.end() ? 0 : it->second.size();
    }
    if (concept) {
        auto it = concept_.find(*concept);
        return it == concept_.end() ? 0 : it->second.size();
    }
    if (feeling) {
        auto it = feeling_.find(*feeling);
        return it == feeling_.end() ? 0 : it->second.size();
    }
    return all_.size();
}

std::map<std::string, size_t> IndexEngine::feeling_distribution() const {
    std::map<std::string, size_t> result;
    for (const auto& [k, v] : feeling_) result[k] = v.size();
    return result;
}

std::map<std::string, std::map<std::string, size_t>>
IndexEngine::concept_feeling_matrix() const {
    std::map<std::string, std::map<std::string, size_t>> matrix;
    for (const auto& [key, v] : cf_)
        matrix[key.first][key.second] = v.size();
    return matrix;
}

// ─── Memory ──────────────────────────────────────────────────────

json Memory::to_json() const {
    return {
        {"memory_id",  memory_id},
        {"concept",    concept},
        {"feeling",    feeling},
        {"media_type", media_type},
        {"content",    content},
        {"note",       note},
        {"tags",       tags},
        {"timestamp",  timestamp},
        {"checksum",   checksum},
    };
}

Memory Memory::from_json(const json& j) {
    Memory m;
    m.memory_id  = j.value("memory_id",  "");
    m.concept    = j.value("concept",    "");
    m.feeling    = j.value("feeling",    "");
    m.media_type = j.value("media_type", "text");
    m.content    = j.value("content",    "");
    m.note       = j.value("note",       "");
    m.timestamp  = j.value("timestamp",  "");
    m.checksum   = j.value("checksum",   "");
    if (j.contains("tags") && j["tags"].is_array())
        m.tags = j["tags"].get<std::vector<std::string>>();
    return m;
}

// ─── MemoryDB ────────────────────────────────────────────────────

MemoryDB::MemoryDB(const std::filesystem::path& path)
    : storage_(path), path_(path.string())
{
    auto records = storage_.scan();
    index_.rebuild(records);
}

std::string MemoryDB::iso_timestamp() {
    auto now = std::chrono::system_clock::now();
    auto t   = std::chrono::system_clock::to_time_t(now);
    std::ostringstream oss;
    oss << std::put_time(std::gmtime(&t), "%FT%TZ");
    return oss.str();
}

std::string MemoryDB::new_uuid() {
    static std::random_device rd;
    static std::mt19937_64 gen(rd());
    std::uniform_int_distribution<uint64_t> dist;
    uint64_t a = dist(gen), b = dist(gen);
    // UUID v4
    a = (a & 0xFFFFFFFFFFFF0FFFULL) | 0x0000000000004000ULL;
    b = (b & 0x3FFFFFFFFFFFFFFFULL) | 0x8000000000000000ULL;
    std::ostringstream oss;
    oss << std::hex << std::setfill('0')
        << std::setw(8) << (a >> 32)        << '-'
        << std::setw(4) << ((a >> 16)&0xFFFF) << '-'
        << std::setw(4) << (a & 0xFFFF)     << '-'
        << std::setw(4) << (b >> 48)        << '-'
        << std::setw(12)<< (b & 0xFFFFFFFFFFFFULL);
    return oss.str();
}

std::string MemoryDB::sha256hex(const std::string& input) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(input.c_str()),
           input.size(), hash);
    std::ostringstream oss;
    for (auto byte : hash)
        oss << std::hex << std::setw(2) << std::setfill('0') << (int)byte;
    return oss.str();
}

Memory MemoryDB::remember(
    const std::string& concept,
    const std::string& feeling,
    const std::string& content,
    const std::string& note,
    const std::vector<std::string>& tags,
    const std::string& media_type)
{
    if (concept.empty()) throw std::invalid_argument("Concept cannot be empty");
    if (content.empty()) throw std::invalid_argument("Content cannot be empty");

    Memory m;
    m.memory_id  = new_uuid();
    m.concept    = concept;
    m.feeling    = feeling;
    m.media_type = media_type;
    m.content    = content;
    m.note       = note;
    m.tags       = tags;
    m.timestamp  = iso_timestamp();
    m.checksum   = sha256hex(content);

    json record = m.to_json();
    uint64_t offset = storage_.append(record);
    index_.index_record(offset, record);
    return m;
}

std::vector<Memory> MemoryDB::recall(
    const std::string& concept,
    const std::string* feeling,
    std::optional<size_t> limit,
    bool oldest_first) const
{
    auto offsets = index_.offsets_by_concept(concept, feeling, oldest_first);
    return load_offsets(offsets, limit);
}

std::vector<Memory> MemoryDB::recall_by_feeling(
    const std::string& feeling,
    std::optional<size_t> limit,
    bool oldest_first) const
{
    auto offsets = index_.offsets_by_feeling(feeling, oldest_first);
    return load_offsets(offsets, limit);
}

std::vector<Memory> MemoryDB::recall_all(
    std::optional<size_t> limit, bool oldest_first) const
{
    auto offsets = index_.all_offsets(oldest_first);
    return load_offsets(offsets, limit);
}

std::vector<Memory> MemoryDB::search(
    const std::string& text,
    bool in_content, bool in_concept, bool in_note,
    std::optional<size_t> limit) const
{
    std::string needle = IndexEngine::to_lower(const_cast<std::string&>(text));
    auto records = storage_.scan();
    std::vector<Memory> results;

    for (auto it = records.rbegin(); it != records.rend(); ++it) {
        const auto& r = it->second;
        bool matched = false;
        auto contains = [&](const std::string& field) {
            std::string val = r.value(field, "");
            std::transform(val.begin(), val.end(), val.begin(), ::tolower);
            return val.find(needle) != std::string::npos;
        };
        if (in_content && contains("content")) matched = true;
        if (in_concept && contains("concept")) matched = true;
        if (in_note    && contains("note"))    matched = true;
        if (matched) {
            results.push_back(record_to_memory(r));
            if (limit && results.size() >= *limit) break;
        }
    }
    return results;
}

size_t MemoryDB::count(const std::string* concept, const std::string* feeling) const {
    return index_.count(concept, feeling);
}

std::map<std::string, size_t> MemoryDB::feeling_distribution() const {
    return index_.feeling_distribution();
}

std::vector<std::string> MemoryDB::list_concepts() const {
    return index_.all_concepts();
}

std::vector<json> MemoryDB::concept_timeline(const std::string& concept) const {
    auto offsets = index_.timeline_offsets(concept);
    std::vector<json> result;
    for (auto offset : offsets) {
        auto r = storage_.read_at(offset);
        if (r) result.push_back({
            {"timestamp", r->value("timestamp", "")},
            {"feeling",   r->value("feeling",   "")},
            {"note",      r->value("note",       "")},
            {"tags",      r->value("tags",       json::array())},
        });
    }
    return result;
}

std::string MemoryDB::export_json() const {
    auto memories = recall_all();
    json arr = json::array();
    for (const auto& m : memories) arr.push_back(m.to_json());
    json out = {{"exported_at", iso_timestamp()}, {"memories", arr}};
    return out.dump(2);
}

std::vector<Memory> MemoryDB::load_offsets(
    const std::vector<uint64_t>& offsets,
    std::optional<size_t> limit) const
{
    std::vector<Memory> result;
    size_t n = limit ? std::min(*limit, offsets.size()) : offsets.size();
    result.reserve(n);
    for (size_t i = 0; i < n; ++i) {
        auto r = storage_.read_at(offsets[i]);
        if (r) result.push_back(record_to_memory(*r));
    }
    return result;
}

Memory MemoryDB::record_to_memory(const json& r) {
    return Memory::from_json(r);
}

} // namespace mnheme

// ─── CMakeLists.txt (minimal) ────────────────────────────────────
// cmake_minimum_required(VERSION 3.20)
// project(mnheme CXX)
// set(CMAKE_CXX_STANDARD 20)
// find_package(OpenSSL REQUIRED)
// find_package(nlohmann_json REQUIRED)
// add_executable(mnheme main.cpp storage.cpp mnheme.cpp)
// target_link_libraries(mnheme PRIVATE OpenSSL::SSL nlohmann_json::nlohmann_json)

File: mnheme_ports/rust/src/mnheme.rs
// mnheme/rust/src/mnheme.rs
// ==========================
// MemoryDB — public API for the MNHEME memory database.
//
// Usage:
//   let mut db = MemoryDB::open("mente.mnheme")?;
//   let mem = db.remember("Debito", Feeling::Ansia, "Ho firmato il mutuo.")?;
//   let memories = db.recall("Debito", None, None, false)?;

use std::io;
use std::path::Path;
use chrono::Utc;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use uuid::Uuid;

use crate::storage::StorageEngine;
use crate::index::IndexEngine;

// ─── Feeling enum ─────────────────────────────────────────────
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Feeling {
    Gioia,
    Tristezza,
    Rabbia,
    Paura,
    Nostalgia,
    Amore,
    Malinconia,
    #[serde(rename = "serenità")]
    Serenita,
    Sorpresa,
    Ansia,
    Gratitudine,
    Vergogna,
    Orgoglio,
    Noia,
    #[serde(rename = "curiosità")]
    Curiosita,
}

impl Feeling {
    pub fn as_str(&self) -> &str {
        match self {
            Feeling::Gioia      => "gioia",
            Feeling::Tristezza  => "tristezza",
            Feeling::Rabbia     => "rabbia",
            Feeling::Paura      => "paura",
            Feeling::Nostalgia  => "nostalgia",
            Feeling::Amore      => "amore",
            Feeling::Malinconia => "malinconia",
            Feeling::Serenita   => "serenità",
            Feeling::Sorpresa   => "sorpresa",
            Feeling::Ansia      => "ansia",
            Feeling::Gratitudine=> "gratitudine",
            Feeling::Vergogna   => "vergogna",
            Feeling::Orgoglio   => "orgoglio",
            Feeling::Noia       => "noia",
            Feeling::Curiosita  => "curiosità",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "gioia"       => Some(Feeling::Gioia),
            "tristezza"   => Some(Feeling::Tristezza),
            "rabbia"      => Some(Feeling::Rabbia),
            "paura"       => Some(Feeling::Paura),
            "nostalgia"   => Some(Feeling::Nostalgia),
            "amore"       => Some(Feeling::Amore),
            "malinconia"  => Some(Feeling::Malinconia),
            "serenità"    => Some(Feeling::Serenita),
            "sorpresa"    => Some(Feeling::Sorpresa),
            "ansia"       => Some(Feeling::Ansia),
            "gratitudine" => Some(Feeling::Gratitudine),
            "vergogna"    => Some(Feeling::Vergogna),
            "orgoglio"    => Some(Feeling::Orgoglio),
            "noia"        => Some(Feeling::Noia),
            "curiosità"   => Some(Feeling::Curiosita),
            _             => None,
        }
    }
}

// ─── Memory struct ────────────────────────────────────────────
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Memory {
    pub memory_id:  String,
    pub concept:    String,
    pub feeling:    String,
    pub media_type: String,
    pub content:    String,
    pub note:       String,
    pub tags:       Vec<String>,
    pub timestamp:  String,
    pub checksum:   String,
}

// ─── Errors ───────────────────────────────────────────────────
#[derive(Debug)]
pub enum MnhemeError {
    Io(io::Error),
    InvalidFeeling(String),
    EmptyConcept,
    EmptyContent,
}

impl From<io::Error> for MnhemeError {
    fn from(e: io::Error) -> Self {
        MnhemeError::Io(e)
    }
}

impl std::fmt::Display for MnhemeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MnhemeError::Io(e)              => write!(f, "IO error: {}", e),
            MnhemeError::InvalidFeeling(s)  => write!(f, "Invalid feeling: {}", s),
            MnhemeError::EmptyConcept       => write!(f, "Concept cannot be empty"),
            MnhemeError::EmptyContent       => write!(f, "Content cannot be empty"),
        }
    }
}

// ─── MemoryDB ─────────────────────────────────────────────────
pub struct MemoryDB {
    storage: StorageEngine,
    index:   IndexEngine,
    path:    String,
}

impl MemoryDB {
    /// Open or create the database at `path`.
    /// Cold start: scans the log and rebuilds indexes.
    pub fn open(path: impl AsRef<Path>) -> Result<Self, MnhemeError> {
        let path_str = path.as_ref().to_string_lossy().to_string();
        let storage  = StorageEngine::new(path.as_ref())?;
        let mut index = IndexEngine::new();
        let records   = storage.scan()?;
        index.rebuild(records);
        Ok(Self { storage, index, path: path_str })
    }

    // ── WRITE ──────────────────────────────────────────────────

    /// Append a new memory. Returns the immutable Memory struct.
    pub fn remember(
        &mut self,
        concept:    &str,
        feeling:    Feeling,
        content:    &str,
        note:       Option<&str>,
        tags:       Option<Vec<String>>,
        media_type: Option<&str>,
    ) -> Result<Memory, MnhemeError> {
        let concept = concept.trim();
        if concept.is_empty() {
            return Err(MnhemeError::EmptyConcept);
        }
        if content.is_empty() {
            return Err(MnhemeError::EmptyContent);
        }

        let memory_id  = Uuid::new_v4().to_string();
        let timestamp  = Utc::now().to_rfc3339();
        let checksum   = format!("{:x}", Sha256::digest(content.as_bytes()));
        let tags_list  = tags.unwrap_or_default();
        let note_str   = note.unwrap_or("");
        let mtype      = media_type.unwrap_or("text");

        let record = json!({
            "memory_id":  memory_id,
            "concept":    concept,
            "feeling":    feeling.as_str(),
            "media_type": mtype,
            "content":    content,
            "note":       note_str,
            "tags":       tags_list,
            "timestamp":  timestamp,
            "checksum":   checksum,
        });

        let offset = self.storage.append(&record)?;
        self.index.index_record(offset, &record);

        Ok(Memory {
            memory_id,
            concept:    concept.to_string(),
            feeling:    feeling.as_str().to_string(),
            media_type: mtype.to_string(),
            content:    content.to_string(),
            note:       note_str.to_string(),
            tags:       tags_list,
            timestamp,
            checksum,
        })
    }

    // ── READ ───────────────────────────────────────────────────

    /// Recall memories by concept, optionally filtered by feeling.
    pub fn recall(
        &self,
        concept:      &str,
        feeling:      Option<&str>,
        limit:        Option<usize>,
        oldest_first: bool,
    ) -> Result<Vec<Memory>, MnhemeError> {
        let offsets = self.index.offsets_by_concept(concept.trim(), feeling, oldest_first);
        self.load_offsets(&offsets, limit)
    }

    /// Recall all memories tagged with a feeling.
    pub fn recall_by_feeling(
        &self,
        feeling:      &str,
        limit:        Option<usize>,
        oldest_first: bool,
    ) -> Result<Vec<Memory>, MnhemeError> {
        let offsets = self.index.offsets_by_feeling(feeling, oldest_first);
        self.load_offsets(&offsets, limit)
    }

    /// Retrieve every memory.
    pub fn recall_all(
        &self,
        limit:        Option<usize>,
        oldest_first: bool,
    ) -> Result<Vec<Memory>, MnhemeError> {
        let offsets = self.index.all_offsets(oldest_first);
        self.load_offsets(&offsets, limit)
    }

    /// Full-text search across content, concept, note.
    pub fn search(
        &self,
        text:       &str,
        in_content: bool,
        in_concept: bool,
        in_note:    bool,
        limit:      Option<usize>,
    ) -> Result<Vec<Memory>, MnhemeError> {
        let needle = text.to_lowercase();
        let mut results = Vec::new();
        let records = self.storage.scan()?;
        for (_, record) in records.iter().rev() {
            let mut matched = false;
            if in_content && record["content"].as_str().unwrap_or("").to_lowercase().contains(&needle) {
                matched = true;
            }
            if in_concept && record["concept"].as_str().unwrap_or("").to_lowercase().contains(&needle) {
                matched = true;
            }
            if in_note && record["note"].as_str().unwrap_or("").to_lowercase().contains(&needle) {
                matched = true;
            }
            if matched {
                results.push(Self::record_to_memory(record));
                if let Some(lim) = limit {
                    if results.len() >= lim {
                        break;
                    }
                }
            }
        }
        Ok(results)
    }

    // ── STATS ──────────────────────────────────────────────────

    pub fn count(&self, concept: Option<&str>, feeling: Option<&str>) -> usize {
        self.index.count(concept, feeling)
    }

    pub fn feeling_distribution(&self) -> std::collections::HashMap<String, usize> {
        let mut dist = self.index.feeling_distribution();
        dist
    }

    pub fn list_concepts(&self) -> Vec<String> {
        self.index.all_concepts()
    }

    pub fn concept_timeline(&self, concept: &str) -> Result<Vec<Value>, MnhemeError> {
        let offsets = self.index.timeline_offsets(concept.trim());
        let mut result = Vec::new();
        for offset in offsets {
            if let Some(r) = self.storage.read_at(offset)? {
                result.push(json!({
                    "timestamp": r["timestamp"],
                    "feeling":   r["feeling"],
                    "note":      r["note"],
                    "tags":      r["tags"],
                }));
            }
        }
        Ok(result)
    }

    // ── EXPORT ────────────────────────────────────────────────

    pub fn export_json(&self) -> Result<String, MnhemeError> {
        let memories = self.recall_all(None, true)?;
        let export = json!({
            "exported_at": Utc::now().to_rfc3339(),
            "memories": memories,
        });
        Ok(serde_json::to_string_pretty(&export)
            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?)
    }

    // ── INTERNAL ──────────────────────────────────────────────

    fn load_offsets(&self, offsets: &[u64], limit: Option<usize>) -> Result<Vec<Memory>, MnhemeError> {
        let iter: Box<dyn Iterator<Item = &u64>> = if let Some(lim) = limit {
            Box::new(offsets.iter().take(lim))
        } else {
            Box::new(offsets.iter())
        };
        let mut result = Vec::new();
        for &offset in iter {
            if let Some(record) = self.storage.read_at(offset)? {
                result.push(Self::record_to_memory(&record));
            }
        }
        Ok(result)
    }

    fn record_to_memory(r: &Value) -> Memory {
        Memory {
            memory_id:  r["memory_id"].as_str().unwrap_or("").to_string(),
            concept:    r["concept"].as_str().unwrap_or("").to_string(),
            feeling:    r["feeling"].as_str().unwrap_or("").to_string(),
            media_type: r["media_type"].as_str().unwrap_or("text").to_string(),
            content:    r["content"].as_str().unwrap_or("").to_string(),
            note:       r["note"].as_str().unwrap_or("").to_string(),
            tags:       r["tags"]
                .as_array()
                .map(|a| a.iter().filter_map(|v| v.as_str().map(String::from)).collect())
                .unwrap_or_default(),
            timestamp:  r["timestamp"].as_str().unwrap_or("").to_string(),
            checksum:   r["checksum"].as_str().unwrap_or("").to_string(),
        }
    }
}

// ─── Cargo.toml ───────────────────────────────────────────────
// [package]
// name    = "mnheme"
// version = "0.1.0"
// edition = "2021"
//
// [dependencies]
// serde       = { version = "1", features = ["derive"] }
// serde_json  = "1"
// chrono      = { version = "0.4", features = ["serde"] }
// uuid        = { version = "1", features = ["v4"] }
// sha2        = "0.10"
// regex       = "1"
